### ISSUE TITLE
Add a way to transform a relative URL in a document

### ISSUE THREAD

**Issue by jasongrout** [jupyter/jupyterlab#446]
Documents may be from different paths, and may have relative URLs to files in their directory, but need a way to request those URLs via http. @minrk and I discussed how we might handle this.

- [X] #504 Implement a javascript function available to a document (perhaps in the context?) to resolve relative URLs. It would take in a relative URL, and return corresponding a full URL to fetch the file from the server (so probably would use something like `http.../baseurl/files/pathtodoc/relativeurl`. This mapping should normalize any `/./` and `/../` and should make sure the path still lies below the baseurl/files/.
- [x] Make a function that takes in a DOM fragment and traverses the DOM fragment looking for things like `<img src=...>`, etc., replacing the relative URLs with appropriate. We'd have to get a good list of places where relative URLs could exist.
- [x] Automatic translation of urls for markdown: this involves hooking into the renderer and translating references. On the other hand, we may just call the above HTML transforming function on the markdown rendered HTML.
- [x] Expose to any user javascript context this relative URL transforming function, probably with the current document path already set in a closure. This probably happens by making the Javascript rendermime a document-specific renderer with this function exposed.

**Comment by jasongrout** [comment]
CC @blink1073

**Comment by blink1073** [comment]
I see the path translation as part of the `contentsmanager` in `js-services`, which will already have the base url.  J-lab users would typically access it through the `ServiceManager`.

**Comment by blink1073** [comment]
```typescript
/**
 * Get a fully qualified url to a file based on its path and an optional base directory.
 */
getUrl(relativePath: string, baseDir = '')
```

**Comment by blink1073** [comment]
The "current" directory comes from the service provider in `src/filebrowser/plugin.ts`.

**Comment by blink1073** [comment]
(If the full path is not known already from a context).

**Comment by jasongrout** [comment]
The URL jupyter-js-services stuff is merged in #504.

**Comment by blink1073** [comment]
Looking at this from the perspective of an `OutputAreaWidget`, it should be able to call `rendermime.render(data, trusted=false)`, and the sanitization and url mangling should happen on the `RenderMime` instance, since the `Rendermime` instance is at the same level as the `DocumentContext`.

- The `RenderMime` instance would have settable `sanitizer` and and `cwd` properties.
- An `IRenderer` would have a `sanitizable` and `safe` attributes, and the bundle would be sanitized before being passed to the renderer if applicable.  A safe output could be chosen over an untrusted mime type of higher preference.
- The `RenderMime` would render the output, then call a function with the `cwd` and the output widget node, which traverses the returned node, looking for `href` and `src` attributes, and mangling as appropriate.

cc @afshin

**Comment by blink1073** [comment]
Note that I have been updating my comment above in-place.

**Comment by blink1073** [comment]
This was fixed in #560.  It turned out that rendering `<img>` tags in the markdown forced us to use the brute-force approach.  A more invasive approach could be considered as a follow-on.
